### PR TITLE
Recreate index schema

### PR DESCRIPTION
**Pull request by mneedham**
This procedure will drop and recreate all indexes and constraints in the database. It's useful if you want to stop using legacy index providers, but don't want to manually drop and create them

Paired on with @JMHReif

### CODE REVIEW

File: src/main/java/apoc/schema/ExceptionUtil.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     * @return the root cause.
+     * @throws IllegalArgumentException if the provided exception is null.
+     */
+    public static Throwable rootCause(Throwable caughtException) {
+        if (null == caughtException) {
+            throw new IllegalArgumentException("Cannot obtain rootCause from (null)");
+        }
+        Throwable root = caughtException;
+        while (root.getCause() != null) {
+            root = root.getCause();
```
Comment (jexp) [inline src/main/java/apoc/schema/ExceptionUtil.java]: cause might be the exception it self so this might end up in an endless loop.

File: src/main/java/apoc/schema/Schemas.java [hunk 2]
```diff
@@ -36,6 +36,23 @@
     @Context
     public KernelTransaction tx;
 
+    @Procedure(value = "apoc.schema.recreate", mode = Mode.SCHEMA)
+    @Description("apoc.schema.recreate() - recreates indexes and constraints")
+    public Stream<AssertSchemaResult> schemaRecreate(@Name(value = "config", defaultValue = "{}") Map<String, Object> config) throws ExecutionException, InterruptedException, TimeoutException {
+        long allIndexesTimeout = (long) config.getOrDefault("retrieveIndexesTimeout", 10000L);
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: the timeouts are bit low
esp. on large graphs those can be hours, so I would increase them massively (which doesn't hurt)

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: where do the current numbers come from?

File: src/main/java/apoc/schema/Schemas.java [hunk 3]
```diff
@@ -36,6 +36,23 @@
...
+        long dropAllIndexesTimeout =  (long) config.getOrDefault("dropAllIndexesTimeout", 30000L);
+        long createAllIndexesTimeout = (long) config.getOrDefault("createAllIndexesTimeout", 30000L);
+
+        long allConstraintsTimeout = (long) config.getOrDefault("retrieveConstraintsTimeout", 10000L);
+        long dropAllConstraintsTimeout =  (long) config.getOrDefault("dropAllConstraintsTimeout", 30000L);
+        long createAllConstraintsTimeout = (long) config.getOrDefault("createAllConstraintsTimeout", 30000L);
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: we can just have a config object instead of individual timeout

File: src/main/java/apoc/schema/Schemas.java [hunk 4]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
         return new AssertSchemaResult(lbl, key).unique().created();
     }
 
+    public List<AssertSchemaResult> recreateIndexes(long allIndexesTimeout, long dropAllIndexesTimeout, long recreateAllIndexesTimeout) throws ExecutionException, InterruptedException, IllegalArgumentException, TimeoutException {
+        Schema schema = db.schema();
+        List<AssertSchemaResult> result = Collections.synchronizedList(new ArrayList<>());
+        ExecutorService executorService = Pools.SINGLE;
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: why not Pools.DEFAULT?

File: src/main/java/apoc/schema/Schemas.java [hunk 5]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
         return new AssertSchemaResult(lbl, key).unique().created();
     }
 
+    public List<AssertSchemaResult> recreateIndexes(long allIndexesTimeout, long dropAllIndexesTimeout, long recreateAllIndexesTimeout) throws ExecutionException, InterruptedException, IllegalArgumentException, TimeoutException {
+        Schema schema = db.schema();
+        List<AssertSchemaResult> result = Collections.synchronizedList(new ArrayList<>());
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: we can just use the results of the future of the 2nd call.
no need for synchronized collection to be passed in

File: src/main/java/apoc/schema/Schemas.java [hunk 6]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
...
+        ExecutorService executorService = Pools.SINGLE;
+        List<IndexDefinition> indexesToRecreate = Util.inTxFuture(executorService, db, () -> {
+            List<IndexDefinition> idx = new ArrayList<>();
+            Iterable<IndexDefinition> outerIndexes = schema.getIndexes();
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: not sure if we need to pull the schema from the db in the separate tx?

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: does it have to be in a separate tx just reading them?

File: src/main/java/apoc/schema/Schemas.java (hunk 2 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: I don't think the operations are synchronous though. The actual command to drop or create should happen straight away I think? And then in the background there would be work going on in an async fashion

File: src/main/java/apoc/schema/Schemas.java (hunk 6 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: It does! This line:

```
schema.getIndexes();
```

Takes a lock on all the indexes and it's only released when the transaction is closed. So you end up in a deadlock if you do that in the procedure's implicit transaction as it would still be holding the lock when the other thread is trying to drop it

File: src/main/java/apoc/schema/ExceptionUtil.java (hunk 1 above)
Comment (mneedham) [reply src/main/java/apoc/schema/ExceptionUtil.java]: Sorry this was added by mistake. Copied across from algos

File: src/main/java/apoc/schema/Schemas.java (hunk 3 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: I'm not sure what a config object is - could you elaborate?

File: src/main/java/apoc/schema/Schemas.java (hunk 4 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: Wanted it to be single threaded. As it is it woudn't make much difference as do blocking futures anyway

File: src/main/java/apoc/schema/Schemas.java (hunk 5 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: Is your suggestion that each call builds up a `List<AssertSchemaResult>` and then we combine them together afterwards?

File: src/main/java/apoc/schema/Schemas.java (hunk 2 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: right for indexes, for constraints it blocks

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: sorry I was referring to db.awaitIndexes() timeouts

File: src/main/java/apoc/schema/Schemas.java (hunk 3 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: just a object that has all those timeouts (encapsulate parameters)

File: src/main/java/apoc/schema/Schemas.java (hunk 4 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: ok

File: src/main/java/apoc/schema/Schemas.java (hunk 5 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: yes

File: src/main/java/apoc/schema/Schemas.java (hunk 6 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: I meant I was not sure if in the separate tx we actually have to call `db.schema()...` there to have the right tx connected to the schema object.

File: src/main/java/apoc/schema/Schemas.java (hunk 3 above)
Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: just a regular java object that we can pass to the methods.

File: src/main/java/apoc/schema/Schemas.java [hunk 7]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
...
+                info.dropped();
+                result.add(info);
+            }
+            return null;
+        }).get(dropAllIndexesTimeout,TimeUnit.MILLISECONDS);
+
+        Util.inTxFuture(executorService, db, () -> {
+            for (IndexDefinition definition : indexesToRecreate) {
+                List<String> propertyKeys = new ArrayList<>();
+                for (String propertyKey : definition.getPropertyKeys()) {
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: little bit inconsistent with above :)

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: And isn't that info already in AssertSchema result? so we can just use that?

File: src/main/java/apoc/schema/Schemas.java [hunk 8]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
...
+            return idx;
+        }).get(allIndexesTimeout, TimeUnit.MILLISECONDS);
+
+        Util.inTxFuture(executorService, db, () -> {
+            for (IndexDefinition definition : indexesToRecreate) {
+                String label = definition.getLabel().name();
+                List<String> keys = new ArrayList<>();
+                definition.getPropertyKeys().forEach(keys::add);
+
+                AssertSchemaResult info = new AssertSchemaResult(label, keys);
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: should also contain the type of index/constraint

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: and compound vs. single

File: src/main/java/apoc/schema/Schemas.java [hunk 9]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
...
+            for (ConstraintDefinition index : outerIndexes) {
+                idx.add(index);
+            }
+            return idx;
+        }).get(allConstraintsTimeout, TimeUnit.MILLISECONDS);
+
+        Util.inTxFuture(executorService, db, () -> {
+            for (ConstraintDefinition constraint : constraintsToRecreate) {
+                String label = constraint.isConstraintType(ConstraintType.RELATIONSHIP_PROPERTY_EXISTENCE) ? constraint.getRelationshipType().name() : constraint.getLabel().name();
+                AssertSchemaResult info = new AssertSchemaResult(label, Iterables.asList(constraint.getPropertyKeys())).unique();
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: third variant of transforming `Iterables.asList(constraint.getPropertyKeys())` into a List :) we should pick one and stay consistent.

File: src/main/java/apoc/schema/Schemas.java [hunk 10]
```diff
@@ -170,7 +283,7 @@ private AssertSchemaResult createSinglePropertyIndex(Schema schema, String lbl,
 
     private AssertSchemaResult createCompoundIndex(String label, List<String> keys) {
         List<String> backTickedKeys = new ArrayList<>();
-        keys.forEach(key->backTickedKeys.add(String.format("`%s`", key)));
+        keys.forEach(key -> backTickedKeys.add(String.format("`%s`", key)));
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: should be just Util.quote

File: src/main/java/apoc/schema/Schemas.java [hunk 11]
```diff
@@ -296,15 +409,15 @@ private Boolean constraintsExistsForRelationship(String type, List<String> prope
...
                 Iterable<ConstraintDescriptor> allConstraints = () -> schemaRead.constraintsGetAll();
-                constraintsIterator = StreamSupport.stream(allConstraints.spliterator(),false)
+                constraintsIterator = StreamSupport.stream(allConstraints.spliterator(), false)
                         .filter(constraint -> Arrays.stream(constraint.schema().getEntityTokenIds()).noneMatch(id -> {
                             try {
                                 return excludeLabels.contains(tokenRead.nodeLabelName(id));
                             } catch (LabelNotFoundKernelException e) {
                                 return false;
                             }
-                            }))
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: lots of unrelated whitespace changes :)

File: src/test/java/apoc/schema/SchemasTest.java [hunk 12]
```diff
@@ -41,6 +48,54 @@ public void tearDown() {
...
+        db.execute("CREATE INDEX ON :Foo(baz)").close();
+
+        List<String> initialIndexNames;
+        try (Transaction tx = db.beginTx()) {
+            initialIndexNames = Iterables.asList(db.schema().getIndexes()).stream().map(Object::toString).collect(Collectors.toList());
+        }
+
+        testResult(db, "CALL apoc.schema.recreate()", (result) -> {
+            while(result.hasNext()) {
+                System.out.println("result.next() = " + result.next());
```
Comment (jexp) [inline src/test/java/apoc/schema/SchemasTest.java]: asserts?

File: src/test/java/apoc/schema/SchemasTest.java [hunk 13]
```diff
@@ -41,6 +48,54 @@ public void tearDown() {
...
+        db.execute("CREATE CONSTRAINT ON (m:Movie) ASSERT m.title IS UNIQUE").close();
+
+        List<String> initialIndexNames;
+        try (Transaction tx = db.beginTx()) {
+            initialIndexNames = Iterables.asList(db.schema().getIndexes()).stream().map(IndexDefinition::getName).collect(Collectors.toList());
+        }
+
+        testResult(db, "CALL apoc.schema.recreate()", (result) -> {
+            while(result.hasNext()) {
+                System.out.println("result.next() = " + result.next());
```
Comment (jexp) [inline src/test/java/apoc/schema/SchemasTest.java]: asserts

File: src/main/java/apoc/schema/Schemas.java (hunk 10 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: We didn't change anything in this function

File: src/main/java/apoc/schema/Schemas.java (hunk 2 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: so for this it's fine as it is?

File: src/main/java/apoc/schema/Schemas.java [hunk 14]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
...
+                constraint.drop();
+                info.dropped();
+                result.add(info);
+            }
+            return null;
+        }).get(dropAllConstraintsTimeout,TimeUnit.MILLISECONDS);
+
+        Util.inTxFuture(executorService, db, () -> {
+            for (ConstraintDefinition constraint : constraintsToRecreate) {
+                List<Object> propertyKeys = new ArrayList<>();
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: consistency around property keys :)

File: src/main/java/apoc/schema/Schemas.java [hunk 15]
```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
         return new AssertSchemaResult(lbl, key).unique().created();
     }
 
+    public List<AssertSchemaResult> recreateIndexes(long allIndexesTimeout, long dropAllIndexesTimeout, long recreateAllIndexesTimeout) throws ExecutionException, InterruptedException, IllegalArgumentException, TimeoutException {
+        Schema schema = db.schema();
+        List<AssertSchemaResult> result = Collections.synchronizedList(new ArrayList<>());
+        ExecutorService executorService = Pools.SINGLE;
+        List<IndexDefinition> indexesToRecreate = Util.inTxFuture(executorService, db, () -> {
+            List<IndexDefinition> idx = new ArrayList<>();
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: can we extract these inner closures into methods so that we can just call them as a method (handle)

or simplify them more into a chain:

1. collect indexes + constraints into AssertResults
2. go over them and drop them
3. go over them and recreate them

File: src/main/java/apoc/schema/Schemas.java (hunk 6 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: I'm not sure what you mean. The reason we have a separate thread to collect all the indexes and constraints is the following code:

File: src/main/java/apoc/schema/Schemas.java (hunk 8 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: This class is copied from the existing `apoc.schema.assert`procedure, we haven't done anything different here

File: src/main/java/apoc/schema/Schemas.java (hunk 14 above)
Comment (mneedham) [reply src/main/java/apoc/schema/Schemas.java]: This one is different because it's a `List<Object>` rather than `List<String>` because that's what the downstream functions take as an argument. So we can't use the Collections approach unless we change the downstream function, and we were trying to change minimal parts of the code.

**Comment by jexp** [comment]
@mneedham what's the state on this?
